### PR TITLE
Soot Spectrum Absorption

### PR DESCRIPTION
**Pull request by owenmylotte**
Incorporates vTCP capability and multiple wavelength transport.

### CODE REVIEW

File: src/eos/radiationProperties/sum.cpp [hunk 1]
```diff
@@ -33,31 +34,102 @@ PetscErrorCode ablate::eos::radiationProperties::Sum::SumTemperatureFunction(con
     PetscFunctionReturn(0);
 }
 
+PetscErrorCode ablate::eos::radiationProperties::Sum::EmissionFunction(const PetscReal *conserved, PetscReal *property, void *ctx) {
```
Comment (mmcgurn) [inline src/eos/radiationProperties/sum.cpp:37]: Are these two functions the same, can we use the same one for both?

File: src/eos/radiationProperties/sum.cpp [hunk 2]
```diff
@@ -33,31 +34,102 @@ PetscErrorCode ablate::eos::radiationProperties::Sum::SumTemperatureFunction(con
...
+    PetscFunctionReturn(0);
+}
+
 ablate::eos::ThermodynamicFunction ablate::eos::radiationProperties::Sum::GetRadiationPropertiesFunction(ablate::eos::radiationProperties::RadiationProperty property,
                                                                                                          const std::vector<domain::Field> &fields) const {
-    // Create the function
-    auto contextVector = std::make_shared<std::vector<ThermodynamicFunction>>();
-    auto function = ThermodynamicFunction{.function = SumFunction, .context = contextVector};
+    switch (property) {
+        case RadiationProperty::Absorptivity: {  // Create the function
```
Comment (mmcgurn) [inline src/eos/radiationProperties/sum.cpp:78]: If we use the same function for both, can we merge this code together?  Should we add code to make sure both property sizes are the same and set the value?

File: CMakeLists.txt [hunk 3]
```diff
@@ -26,7 +26,7 @@ if (CMAKE_PROJECT_NAME STREQUAL PROJECT_NAME)
 endif ()
 
 # Create the library for ablate
-add_library(ablateLibrary SHARED)
+add_library(ablateLibrary SHARED src/eos/radiationProperties/sootSpectrumProperties.cpp src/eos/radiationProperties/sootSpectrumProperties.hpp)
```
Comment (mmcgurn) [inline CMakeLists.txt:29]: Can you moves this to the local make file?

File: src/eos/radiationProperties/constant.hpp [hunk 4]
```diff
@@ -1,32 +1,66 @@
...
+
+    const std::shared_ptr<eos::EOS> eos;  //! eos is needed to compute temperature
+    PetscReal absorptivityIn;
+    PetscReal emissivityIn;
+
+    /**
+     * Get the refractive index for the material.
+     * @return
+     */
+    static inline PetscReal GetRefractiveIndex() { return 1; }
```
Comment (mmcgurn) [inline src/eos/radiationProperties/constant.hpp:25]: You may want to add constexpr to this function

File: src/eos/radiationProperties/constant.cpp [hunk 5]
```diff
@@ -1,22 +1,54 @@
...
+    PetscReal refractiveIndex = GetRefractiveIndex();  //! We may want to incorporate this at some point?
+    *(property) = functionContext->emissivity * ablate::radiation::Radiation::GetBlackBodyTotalIntensity(temperature, refractiveIndex);
+    PetscFunctionReturn(0);
+}
+
+PetscErrorCode ablate::eos::radiationProperties::Constant::ConstantEmissionTemperatureFunction(const PetscReal conserved[], PetscReal temperature, PetscReal *property, void *ctx) {
+    PetscFunctionBeginUser;
+    auto functionContext = (FunctionContext *)ctx;
+    PetscReal refractiveIndex = GetRefractiveIndex();  //! We may want to incorporate this at some point?
+    *(property) = functionContext->emissivity * ablate::radiation::Radiation::GetBlackBodyTotalIntensity(temperature, refractiveIndex);
```
Comment (mmcgurn) [inline src/eos/radiationProperties/constant.cpp:36]: What is being returned from this function?  Why can't we compute the Emissivity and multiply by the total intensity elsewhere?

File: src/io/hdf5MultiFileSerializer.cpp [hunk 6]
```diff
@@ -211,6 +211,34 @@ std::filesystem::path ablate::io::Hdf5MultiFileSerializer::GetOutputFilePath(con
 
 std::filesystem::path ablate::io::Hdf5MultiFileSerializer::GetOutputDirectoryPath(const std::string& objectId) { return environment::RunEnvironment::Get().GetOutputDirectory() / objectId; }
 
+void ablate::io::Hdf5MultiFileSerializer::RestoreFromSequence(PetscInt currentSequenceNumber, std::weak_ptr<Serializable> serializable) {
```
Comment (mmcgurn) [inline src/io/hdf5MultiFileSerializer.cpp:214]: Is this code working?

File: src/io/serializer.hpp [hunk 7]
```diff
@@ -18,6 +19,11 @@ class Serializer {
     virtual void* GetContext() { return this; }
     virtual PetscSerializeFunction GetSerializeFunction() = 0;
     virtual void RestoreTS(TS ts) = 0;
+
```
Comment (mmcgurn) [inline src/io/serializer.hpp:22]: Can you add some doxygen comment to this to explain why it is here.  Also, I think that we may wan to use the operator overloading ()

File: src/monitors/monitor.hpp [hunk 8]
```diff
@@ -31,6 +31,11 @@ class Monitor {
      */
     virtual void* GetContext() { return this; }
 
+    void CallMonitor(TS ts, PetscInt steps, PetscReal time, Vec u) {
```
Comment (mmcgurn) [inline src/monitors/monitor.hpp:34]: I think that we may wan to use the operator overloading ()

File: src/radiation/radiation.cpp [hunk 9]
```diff
@@ -569,34 +560,48 @@ void ablate::radiation::Radiation::EvaluateGains(Vec solVec, ablate::domain::Fie
...
 
     // Start by marching over all rays in this rank
-    for (std::size_t r = 0; r < raySegments.size(); r++) {
-        // zero our the calculation
-        auto& raySegmentsCalculation = raySegmentsCalculations[r];
-        raySegmentsCalculation.Ij = 0.0;
-        raySegmentsCalculation.Krad = 1.0;
+    for (std::size_t raySegmentIndex = 0; raySegmentIndex < raySegments.size(); ++raySegmentIndex) {
+        //! Zero this ray segment for all wavelengths
+        for (unsigned short int wavelengthIndex = 0; wavelengthIndex < absorptivityFunction.propertySize; wavelengthIndex++) {  //! Iterate through every wavelength entry in this ray segment
```
Comment (mmcgurn) [inline src/radiation/radiation.cpp:568]: What type is the propertySize?

File: src/radiation/radiation.hpp [hunk 10]
```diff
@@ -4,13 +4,23 @@
 #include <memory>
 #include <set>
 #include "eos/radiationProperties/radiationProperties.hpp"
+#include "eos/radiationProperties/sootSpectrumProperties.hpp"
```
Comment (mmcgurn) [inline src/radiation/radiation.hpp:7]: Why does radiation.hpp need to know about soot properties?

File: tests/integrationTests/inputs/radiation/spectrumRadiation.yaml [hunk 11]
```diff
@@ -0,0 +1,89 @@
+#
```
Comment (mmcgurn) [inline tests/integrationTests/inputs/radiation/spectrumRadiation.yaml:1]: Can you add some documentation to this example that explains why it is being used and what it does?

File: tests/integrationTests/inputs/radiation/virtualTCP.yaml [hunk 12]
```diff
@@ -0,0 +1,104 @@
+#
```
Comment (mmcgurn) [inline tests/integrationTests/inputs/radiation/virtualTCP.yaml:1]: Same here please?

File: tests/integrationTests/outputs/radiation/parallelPlatesOutput.txt [hunk 13]
```diff
@@ -1,21 +1,21 @@
 MinMaxAvg euler for timestep 0000:
-	min: [0.177, 2.53e+05,  0,  0]
-	max: [2.75, 2.53e+05,  0,  0]
-	avg: [0.6, 2.53e+05,  0,  0]
+min: \[(.*), (.*), (.*), (.*)\]<expects> ~ ~ ~ ~
```
Comment (mmcgurn) [inline tests/integrationTests/outputs/radiation/parallelPlatesOutput.txt:2]: Can you add some sanity checks for this numbers?

**Review by mmcgurn — CHANGES_REQUESTED** [review]

File: src/radiation/radiation.cpp (hunk 9 above)
Comment (owenmylotte) [reply src/radiation/radiation.cpp:568]: It is a PetscInt. Maybe this can be a smaller data type?

Comment (mmcgurn) [reply src/radiation/radiation.cpp:568]: You probably want to match absorptivityFunction.propertySize with unsigned short int.  I think that propertySize is a PetscInt.  PetscInt type changes depending upon how petsc is built.

**Review by mmcgurn — CHANGES_REQUESTED** [review]
Thanks for updating this, just a few small things.

**Review by mmcgurn — CHANGES_REQUESTED** [review]

**Review by mmcgurn — APPROVED** [review]

**Pull request merged**